<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>scNMTseq challenge analysis using a PLS-based approach • BIRSBIO2020.scNMTseq.PLS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="scNMTseq challenge analysis using a PLS-based approach">
<meta property="og:description" content="BIRSBIO2020.scNMTseq.PLS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BIRSBIO2020.scNMTseq.PLS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/index.html">scNMTseq challenge analysis using a PLS-based approach</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ajabadi/BIRSBIO2020.scNMTseq.PLS">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="index_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>scNMTseq challenge analysis using a PLS-based approach</h1>
                        <h4 class="author">Al JalalAbadi<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</h4>
            
            <h4 class="date">22 Jul 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ajabadi/BIRSBIO2020.scNMTseq.PLS/blob/master/vignettes/index.Rmd"><code>vignettes/index.Rmd</code></a></small>
      <div class="hidden name"><code>index.Rmd</code></div>

    </div>

    
    
<p>To Do</p>
<ul>
<li>Add kNN</li>
</ul>
<hr>
<p><strong>Note:</strong> These analyses are (slightly) different from the ones presented in the hackathon in the sense that:</p>
<ul>
<li>The cells used are different (only embryonic cells used)</li>
<li>To enhance reproducibility on different machines, <code><a href="https://rdrr.io/pkg/uwot/man/umap.html">uwot::umap</a></code> was used instead of <code><a href="https://rdrr.io/pkg/umap/man/umap.html">umap::umap</a></code> for dimensionality reduction of RNA data so that python dependencies are reduced</li>
</ul>
<hr>
<p>Load the required packages:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BIRSBIO2020.scNMTseq.PLS</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">MultiAssayExperiment</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">mixOmics</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">knitr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">uwot</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">nipals</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">biomaRt</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">magrittr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">reshape2</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">data.table</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">MOFA2</span>)</pre></body></html></div>
<div id="data" class="section level1">
<h1 class="hasAnchor">
<a href="#data" class="anchor"></a>Data</h1>
<p>Details of the hackathon data and preprocessing steps: <a href="https://github.com/BIRSBiointegration/Hackathon/tree/master/scNMT-seq" class="uri">https://github.com/BIRSBiointegration/Hackathon/tree/master/scNMT-seq</a></p>
<p>Load the <code>MultiAssayExperiment</code> object:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span>(<span class="st">'loading data from %s ...\n'</span>, <span class="no">params</span>$<span class="no">data_path</span>))
<span class="co">#&gt; loading data from https://cloudstor.aarnet.edu.au/plus/s/Xzf5vCgAEUVgbfQ/download?path=%2Foutput&amp;files=scnmtseq_gastrulation_mae_826-cells_orderedFeatures.rds ...</span>
<span class="no">gastru.mae</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/connections.html">url</a></span>(<span class="no">params</span>$<span class="no">data_path</span>))</pre></body></html></div>
<p>An overview of the data object:</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">gastru.mae</span>
<span class="co">#&gt; A MultiAssayExperiment object of 13 listed</span>
<span class="co">#&gt;  experiments with user-defined names and respective classes.</span>
<span class="co">#&gt;  Containing an ExperimentList class object of length 13:</span>
<span class="co">#&gt;  [1] rna: matrix with 18345 rows and 826 columns</span>
<span class="co">#&gt;  [2] met_genebody: matrix with 15837 rows and 826 columns</span>
<span class="co">#&gt;  [3] met_promoter: matrix with 12092 rows and 826 columns</span>
<span class="co">#&gt;  [4] met_cgi: matrix with 5536 rows and 826 columns</span>
<span class="co">#&gt;  [5] met_p300: matrix with 101 rows and 826 columns</span>
<span class="co">#&gt;  [6] met_CTCF: matrix with 175 rows and 826 columns</span>
<span class="co">#&gt;  [7] met_DHS: matrix with 66 rows and 826 columns</span>
<span class="co">#&gt;  [8] acc_genebody: matrix with 17139 rows and 826 columns</span>
<span class="co">#&gt;  [9] acc_promoter: matrix with 16518 rows and 826 columns</span>
<span class="co">#&gt;  [10] acc_cgi: matrix with 4459 rows and 826 columns</span>
<span class="co">#&gt;  [11] acc_p300: matrix with 138 rows and 826 columns</span>
<span class="co">#&gt;  [12] acc_CTCF: matrix with 898 rows and 826 columns</span>
<span class="co">#&gt;  [13] acc_DHS: matrix with 290 rows and 826 columns</span>
<span class="co">#&gt; Functionality:</span>
<span class="co">#&gt;  experiments() - obtain the ExperimentList instance</span>
<span class="co">#&gt;  colData() - the primary/phenotype DataFrame</span>
<span class="co">#&gt;  sampleMap() - the sample coordination DataFrame</span>
<span class="co">#&gt;  `$`, `[`, `[[` - extract colData columns, subset, or experiment</span>
<span class="co">#&gt;  *Format() - convert into a long or wide DataFrame</span>
<span class="co">#&gt;  assays() - convert ExperimentList to a SimpleList of matrices</span>
<span class="co">#&gt;  exportClass() - save all data to files</span></pre></body></html></div>
<div id="filter-extra-embryonic-cells" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-extra-embryonic-cells" class="anchor"></a>filter extra-embryonic cells</h2>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span>(<span class="st">'Dropping cells from lineage(s) %s:\n'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sQuote.html">sQuote</a></span>(<span class="no">params</span>$<span class="no">drop_lineages</span>), <span class="kw">collapse</span> <span class="kw">=</span> <span class="st">', '</span>)))
<span class="co">#&gt; Dropping cells from lineage(s) 'Visceral_endoderm', 'ExE_ectoderm':</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span>(<span class="st">'Retaining assays: %s: \n'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sQuote.html">sQuote</a></span>(<span class="no">params</span>$<span class="no">keep_assays</span>), <span class="kw">collapse</span> <span class="kw">=</span> <span class="st">', '</span>)))
<span class="co">#&gt; Retaining assays: 'rna', 'met_genebody', 'met_promoter', 'met_cgi', 'met_p300', 'met_CTCF', 'met_DHS':</span>
<span class="no">gastru.mae</span> <span class="kw">&lt;-</span> <span class="no">gastru.mae</span>[, !(<span class="no">gastru.mae</span>$<span class="no">lineage10x_2</span> <span class="kw">%in%</span> <span class="no">params</span>$<span class="no">drop_lineages</span>), <span class="no">params</span>$<span class="no">keep_assays</span>]
<span class="co">#&gt; harmonizing input:</span>
<span class="co">#&gt;   removing 4494 sampleMap rows not in names(experiments)</span></pre></body></html></div>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">gastru.mae</span>
<span class="co">#&gt; A MultiAssayExperiment object of 7 listed</span>
<span class="co">#&gt;  experiments with user-defined names and respective classes.</span>
<span class="co">#&gt;  Containing an ExperimentList class object of length 7:</span>
<span class="co">#&gt;  [1] rna: matrix with 18345 rows and 749 columns</span>
<span class="co">#&gt;  [2] met_genebody: matrix with 15837 rows and 749 columns</span>
<span class="co">#&gt;  [3] met_promoter: matrix with 12092 rows and 749 columns</span>
<span class="co">#&gt;  [4] met_cgi: matrix with 5536 rows and 749 columns</span>
<span class="co">#&gt;  [5] met_p300: matrix with 101 rows and 749 columns</span>
<span class="co">#&gt;  [6] met_CTCF: matrix with 175 rows and 749 columns</span>
<span class="co">#&gt;  [7] met_DHS: matrix with 66 rows and 749 columns</span>
<span class="co">#&gt; Functionality:</span>
<span class="co">#&gt;  experiments() - obtain the ExperimentList instance</span>
<span class="co">#&gt;  colData() - the primary/phenotype DataFrame</span>
<span class="co">#&gt;  sampleMap() - the sample coordination DataFrame</span>
<span class="co">#&gt;  `$`, `[`, `[[` - extract colData columns, subset, or experiment</span>
<span class="co">#&gt;  *Format() - convert into a long or wide DataFrame</span>
<span class="co">#&gt;  assays() - convert ExperimentList to a SimpleList of matrices</span>
<span class="co">#&gt;  exportClass() - save all data to files</span></pre></body></html></div>
<p>Breakdown of the cells in each stage:</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">gastru.mae</span>$<span class="no">stage</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/magrittr/man/aliases.html">set_colnames</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'stage'</span>, <span class="st">'# of cells'</span>)) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span>()</pre></body></html></div>
<table class="table">
<thead><tr class="header">
<th align="left">stage</th>
<th align="right"># of cells</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">E4.5</td>
<td align="right">104</td>
</tr>
<tr class="even">
<td align="left">E5.5</td>
<td align="right">84</td>
</tr>
<tr class="odd">
<td align="left">E6.5</td>
<td align="right">218</td>
</tr>
<tr class="even">
<td align="left">E7.5</td>
<td align="right">343</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="co">## create a data.frame from cell metadata</span>
<span class="no">coldata</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="fu">colData</span>(<span class="no">gastru.mae</span>))</pre></body></html></div>
</div>
<div id="feature-detection" class="section level2">
<h2 class="hasAnchor">
<a href="#feature-detection" class="anchor"></a>Feature detection</h2>
<p>Create density plots of the feature detection rate across all cells for all modalities:</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="co"># get the methylation assays</span>
<span class="no">met_assays</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">gastru.mae</span>), <span class="kw">pattern</span> <span class="kw">=</span> <span class="st">'^met'</span>, <span class="kw">value</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="co"># add dimensions to labels for ggplot</span>
<span class="no">dims</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/MultiAssayExperiment/man/MultiAssayExperiment-methods.html">experiments</a></span>(<span class="no">gastru.mae</span>[,,<span class="no">met_assays</span>]), <span class="no">dim</span>)
<span class="co">#&gt; harmonizing input:</span>
<span class="co">#&gt;   removing 749 sampleMap rows not in names(experiments)</span>
<span class="no">dims</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="no">dims</span>, <span class="kw">function</span>(<span class="no">x</span>) <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span>(<span class="st">' (%s, %s)'</span>, <span class="no">x</span>[<span class="fl">2</span>], <span class="no">x</span>[<span class="fl">1</span>]))
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">met_assays</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="no">met_assays</span>, <span class="no">dims</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">as.list</a></span>()
<span class="no">met_assays</span>
<span class="co">#&gt; met_genebody (749, 15837) met_promoter (749, 12092)       met_cgi (749, 5536) </span>
<span class="co">#&gt;            "met_genebody"            "met_promoter"                 "met_cgi" </span>
<span class="co">#&gt;       met_p300 (749, 101)       met_CTCF (749, 175)         met_DHS (749, 66) </span>
<span class="co">#&gt;                "met_p300"                "met_CTCF"                 "met_DHS"</span>
<span class="co"># calculate the feature detection in a data.frame for methylation assays</span>
<span class="no">coverages</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">met_assays</span>, <span class="kw">function</span>(<span class="no">assay_name</span>) {
  <span class="no">mat</span> <span class="kw">&lt;-</span> <span class="fu">assay</span>(<span class="no">gastru.mae</span>, <span class="no">assay_name</span>)
  <span class="no">NAs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(!<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">mat</span>))/<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">mat</span>)[<span class="fl">2</span>]*<span class="fl">100</span>
  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">pct_NAs</span><span class="kw">=</span><span class="no">NAs</span>)
})
<span class="co"># create a long data.frame containing the assay name for plot</span>
<span class="no">coverages</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://github.com/ajabadi/BIRSBIO2020.scNMTseq.PLS/reference/utils.html">rbindListWithNames</a></span>(<span class="no">coverages</span>)
<span class="no">coverages</span>$<span class="no">dataset</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">coverages</span>$<span class="no">dataset</span>, <span class="kw">levels</span> <span class="kw">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/duplicated.html">unique</a></span>(<span class="no">coverages</span>$<span class="no">dataset</span>), <span class="kw">ordered</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<p>Genomic contexts vary in level of feature detection rates:</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">cov_plot</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">coverages</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">pct_NAs</span>)) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span>(<span class="kw">fill</span> <span class="kw">=</span> <span class="st">'lightblue'</span>, <span class="kw">show.legend</span> <span class="kw">=</span> <span class="fl">FALSE</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">xintercept</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(<span class="no">pct_NAs</span>)),
             <span class="kw">color</span><span class="kw">=</span><span class="st">"blue"</span>, <span class="kw">linetype</span><span class="kw">=</span><span class="st">"dashed"</span>, <span class="kw">size</span><span class="kw">=</span><span class="fl">0.5</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="st">'% of cells detecting the feature'</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="no">.</span>~<span class="no">dataset</span>, <span class="kw">nrow</span> <span class="kw">=</span> <span class="fl">2</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="kw">strip.text.x</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="kw">size</span> <span class="kw">=</span> <span class="fl">10</span>, <span class="kw">face</span> <span class="kw">=</span> <span class="st">'bold'</span>, <span class="kw">color</span> <span class="kw">=</span> <span class="st">'purple'</span>))

<span class="no">cov_plot</span></pre></body></html></div>
<div class="figure">
<img src="index_files/figure-html/unnamed-chunk-12-1.png" alt="density plots for methylation data show that shorter genomic regions tend to have less feture coverage. Dashed blue line indicates the average across all modalities." width="768"><p class="caption">
density plots for methylation data show that shorter genomic regions tend to have less feture coverage. Dashed blue line indicates the average across all modalities.
</p>
</div>
</div>
</div>
<div id="rna" class="section level1">
<h1 class="hasAnchor">
<a href="#rna" class="anchor"></a>RNA</h1>
<div id="umap" class="section level2">
<h2 class="hasAnchor">
<a href="#umap" class="anchor"></a>UMAP</h2>
<p>Here we use <code>uwot</code> package for Uniform Manifold Approximation Projection. The original study used the <code>umap</code> package with <code>method = 'umap-learn'</code> parameter which requires the ‘umap-learn’ python library and can add complexity to set up on a VM.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="no">params</span>$<span class="no">umap_params</span>[<span class="st">'run.seed'</span>])
<span class="co">## all params use defaults except for 'n_components' and 'min_dist'</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span>(<span class="st">'running umap on RNA data with parameters: %s ...\n '</span>,
            <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">params</span>$<span class="no">umap_params</span>[-<span class="fl">1</span>]), <span class="st">':'</span>,
                  <span class="no">params</span>$<span class="no">umap_params</span>[-<span class="fl">1</span>], <span class="kw">collapse</span> <span class="kw">=</span> <span class="st">', '</span>)))
<span class="co">#&gt; running umap on RNA data with parameters: n_neighbors : 15, n_components : 4 ...</span>
<span class="co">#&gt; </span>
<span class="no">umap_rna_unsup</span> <span class="kw">&lt;-</span> <span class="kw pkg">uwot</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/uwot/man/umap.html">umap</a></span>(
  <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="fu">assay</span>(<span class="no">gastru.mae</span>, <span class="st">'rna'</span>)),
  <span class="kw">n_neighbors</span> <span class="kw">=</span> <span class="no">params</span>$<span class="no">umap_params</span>[<span class="st">'n_neighbors'</span>],
  <span class="kw">n_components</span> <span class="kw">=</span> <span class="no">params</span>$<span class="no">umap_params</span>[<span class="st">'n_components'</span>],
  <span class="kw">metric</span> <span class="kw">=</span> <span class="st">"euclidean"</span>,
  <span class="kw">n_epochs</span> <span class="kw">=</span> <span class="kw">NULL</span>,
  <span class="kw">learning_rate</span> <span class="kw">=</span> <span class="fl">1</span>,
  <span class="kw">scale</span> <span class="kw">=</span> <span class="fl">FALSE</span>,
  <span class="kw">init</span> <span class="kw">=</span> <span class="st">"spectral"</span>,
  <span class="kw">init_sdev</span> <span class="kw">=</span> <span class="kw">NULL</span>,
  <span class="kw">spread</span> <span class="kw">=</span> <span class="fl">1</span>,
  <span class="kw">set_op_mix_ratio</span> <span class="kw">=</span> <span class="fl">1</span>,
  <span class="kw">local_connectivity</span> <span class="kw">=</span> <span class="fl">1</span>)</pre></body></html></div>
<p>Create a data.frame of embeddings:</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">df</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="no">umap_rna_unsup</span>)
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">df</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">'UMAP_'</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span>(<span class="no">df</span>))
<span class="no">df</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="no">df</span>, <span class="no">coldata</span>)</pre></body></html></div>
<p>Create colour palettes for stages and lineages:</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="co"># col_pallete &lt;- viridisLite::viridis(n = 4)</span>
<span class="no">col_pallete</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#440154FF"</span>, <span class="st">"#31688EFF"</span>, <span class="st">"#35B779FF"</span>, <span class="st">"#FDE725FF"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">col_pallete</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"E4.5"</span>, <span class="st">"E5.5"</span>, <span class="st">"E6.5"</span>, <span class="st">"E7.5"</span>)</pre></body></html></div>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="co">## helper function to create both plots coloured by stage</span>
<span class="no">plot_umap_by_stage</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">df</span>, <span class="no">dims</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">2</span>)) {
  <span class="no">axes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">'UMAP_'</span>, <span class="no">dims</span>)
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes_.html">aes_string</a></span>(<span class="no">axes</span>[<span class="fl">1</span>], <span class="no">axes</span>[<span class="fl">2</span>])) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">col</span><span class="kw">=</span><span class="no">stage</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span>()+ <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="no">col_pallete</span>)
}</pre></body></html></div>
<p>First two components:</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="fu">plot_umap_by_stage</span>(<span class="kw">df</span> <span class="kw">=</span> <span class="no">df</span>, <span class="kw">dims</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">2</span>))</pre></body></html></div>
<div class="figure">
<img src="index_files/figure-html/unnamed-chunk-19-1.png" alt="UMAP plot shows distinct cell populations corresponding to early-stage, mid-stage, and late-stage cells as well as the within-stage variation for mainly E4.5 and E6.5 which could correspond to different lineages." width="672"><p class="caption">
UMAP plot shows distinct cell populations corresponding to early-stage, mid-stage, and late-stage cells as well as the within-stage variation for mainly E4.5 and E6.5 which could correspond to different lineages.
</p>
</div>
</div>
</div>
<div id="integration" class="section level1">
<h1 class="hasAnchor">
<a href="#integration" class="anchor"></a>Integration</h1>
<div id="gene-id-to-gene-symbol-conversion-data-frame" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-id-to-gene-symbol-conversion-data-frame" class="anchor"></a>gene id to gene symbol conversion data.frame</h2>
<p>Get all the gene names from all modalities:</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="no">all_features</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/MultiAssayExperiment/man/MultiAssayExperiment-methods.html">experiments</a></span>(<span class="no">gastru.mae</span>), <span class="no">rownames</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>()
<span class="no">all_genes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="kw">pattern</span> <span class="kw">=</span> <span class="st">'^ENSMUSG'</span>, <span class="kw">x</span> <span class="kw">=</span> <span class="no">all_features</span>, <span class="kw">value</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="no">all_genes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/duplicated.html">unique</a></span>(<span class="no">all_genes</span>)
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">all_genes</span>)
<span class="co">#&gt; [1] 18359</span></pre></body></html></div>
<p>Get the Ensembl gene name for these gene ids:</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">ensembl</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html">useMart</a></span>(<span class="st">"ensembl"</span>, <span class="kw">dataset</span><span class="kw">=</span><span class="st">"mmusculus_gene_ensembl"</span>)
<span class="no">mouse_gene_ids</span>  <span class="kw">&lt;-</span> <span class="no">all_genes</span>

<span class="no">symbols</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span>(<span class="kw">attributes</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'ensembl_gene_id'</span>,
                          <span class="st">'external_gene_name'</span>),
             <span class="kw">filters</span> <span class="kw">=</span> <span class="st">"ensembl_gene_id"</span>,
             <span class="kw">values</span> <span class="kw">=</span> <span class="no">all_genes</span>,
             <span class="kw">mart</span> <span class="kw">=</span> <span class="no">ensembl</span>)
<span class="no">symbols</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="no">symbols</span>, <span class="kw">row.names</span> <span class="kw">=</span> <span class="fl">1</span>)</pre></body></html></div>
</div>
<div id="pls-integartion-of-all-modalities-with-feature-selection" class="section level2">
<h2 class="hasAnchor">
<a href="#pls-integartion-of-all-modalities-with-feature-selection" class="anchor"></a>PLS integartion of all modalities with feature selection</h2>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="co"># data dimensions</span>
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/MultiAssayExperiment/man/MultiAssayExperiment-methods.html">experiments</a></span>(<span class="no">gastru.mae</span>), <span class="no">dim</span>)
<span class="co">#&gt; $rna</span>
<span class="co">#&gt; [1] 18345   749</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $met_genebody</span>
<span class="co">#&gt; [1] 15837   749</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $met_promoter</span>
<span class="co">#&gt; [1] 12092   749</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $met_cgi</span>
<span class="co">#&gt; [1] 5536  749</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $met_p300</span>
<span class="co">#&gt; [1] 101 749</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $met_CTCF</span>
<span class="co">#&gt; [1] 175 749</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $met_DHS</span>
<span class="co">#&gt; [1]  66 749</span></pre></body></html></div>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="no">mmspls</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://github.com/ajabadi/BIRSBIO2020.scNMTseq.PLS/reference/multimodal_analysis_wrapper.html">multimodal_analysis_wrapper</a></span>(<span class="kw">mae</span> <span class="kw">=</span> <span class="no">gastru.mae</span>, <span class="kw">study_assays</span> <span class="kw">=</span> <span class="kw">NULL</span>, <span class="kw">ncomp</span> <span class="kw">=</span> <span class="fl">2</span>, <span class="kw">scale</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">design</span> <span class="kw">=</span> <span class="st">'null'</span>, <span class="kw">lineages</span> <span class="kw">=</span> <span class="kw">NULL</span>, <span class="kw">stages</span> <span class="kw">=</span> <span class="kw">NULL</span>, <span class="kw">DA</span> <span class="kw">=</span> <span class="kw">NULL</span>, <span class="kw">keepX</span> <span class="kw">=</span> <span class="kw">NULL</span>, <span class="kw">save</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="co">#&gt; Making unique feature names by adding assay name to feature name for: met_genebody</span>
<span class="co">#&gt; Making unique feature names by adding assay name to feature name for: met_promoter</span>
<span class="co">#&gt; Warning in cor(A[[k]], variates.A[[k]]): the standard deviation is zero</span></pre></body></html></div>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="no">plotIndiv_wrapper</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">pls_obj</span>, <span class="no">coldata</span>, <span class="no">col_cell</span> <span class="kw">=</span> <span class="st">'stage'</span>, <span class="no">legend.title</span> <span class="kw">=</span> <span class="st">'Stage'</span>, <span class="no">comp</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">2</span>), <span class="no">...</span>) {
  <span class="co">## order coldata cells based on matching cells in pls modalities</span>
  <span class="no">coldata</span> <span class="kw">&lt;-</span> <span class="no">coldata</span>[<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">pls_obj</span>$<span class="no">X</span><span class="kw">[[</span><span class="fl">1</span>]]),]
  <span class="no">cell_class</span> <span class="kw">&lt;-</span> <span class="no">coldata</span>[,<span class="no">col_cell</span>]  <span class="co">## col factor for cells</span>
  <span class="no">cell_class</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">cell_class</span>)
  <span class="fu"><a href="https://rdrr.io/r/base/warning.html">suppressWarnings</a></span>({
    <span class="fu"><a href="https://rdrr.io/pkg/mixOmics/man/plotIndiv.html">plotIndiv</a></span>(<span class="no">pls_obj</span>, <span class="kw">pch</span> <span class="kw">=</span> <span class="fl">16</span>, <span class="kw">comp</span> <span class="kw">=</span> <span class="no">comp</span>,<span class="kw">group</span> <span class="kw">=</span> <span class="no">cell_class</span>, <span class="kw">legend</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">legend.title</span> <span class="kw">=</span> <span class="no">legend.title</span>, <span class="kw">size.subtitle</span> <span class="kw">=</span> <span class="fl">10</span>, <span class="kw">cex</span><span class="kw">=</span><span class="fl">0.8</span>, <span class="no">...</span>)
  })
}</pre></body></html></div>
</div>
<div id="score-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#score-plots" class="anchor"></a>score plots</h2>
<p>Colored by stage:</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="fu">plotIndiv_wrapper</span>(<span class="kw">pls_obj</span> <span class="kw">=</span> <span class="no">mmspls</span>, <span class="kw">coldata</span> <span class="kw">=</span> <span class="no">coldata</span>, <span class="kw">col_cell</span> <span class="kw">=</span> <span class="st">'stage'</span>, <span class="kw">legend.title</span> <span class="kw">=</span> <span class="st">'Stage'</span>, <span class="kw">comp</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">2</span>), <span class="kw">subtitle</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">mmspls</span>$<span class="no">X</span>), <span class="kw">col.per.group</span> <span class="kw">=</span> <span class="no">col_pallete</span>)</pre></body></html></div>
<p><img src="index_files/figure-html/unnamed-chunk-27-1.png" width="768"></p>
</div>
</div>
<div id="concordancy-with-rna" class="section level1">
<h1 class="hasAnchor">
<a href="#concordancy-with-rna" class="anchor"></a>Concordancy with RNA</h1>
<p>Average correlation of the components of rna and different modalities:</p>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="no">cor_with_rna</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="no">mmspls</span>$<span class="no">variates</span>[-<span class="fl">1</span>], <span class="kw">function</span>(<span class="no">x</span>) {
  <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span>(<span class="no">mmspls</span>$<span class="no">variates</span>$<span class="no">rna</span>, <span class="no">x</span>)))
})

<span class="no">cor_with_rna</span> <span class="kw">&lt;-</span> <span class="no">cor_with_rna</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="kw">decreasing</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="no">cor_with_rna</span> <span class="kw">&lt;-</span> <span class="no">cor_with_rna</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/magrittr/man/aliases.html">set_colnames</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'cor'</span>))
<span class="no">cor_with_rna</span>$<span class="no">modality</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">cor_with_rna</span>), <span class="kw">levels</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">cor_with_rna</span>), <span class="kw">ordered</span> <span class="kw">=</span> <span class="fl">TRUE</span>)

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">cor_with_rna</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">modality</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">cor</span>)) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_col</a></span>(<span class="kw">fill</span> <span class="kw">=</span> <span class="st">"#35B779FF"</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span>() +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">y</span> <span class="kw">=</span> <span class="st">'Avg correlation with RNA'</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="st">'Modality'</span>)</pre></body></html></div>
<p><img src="index_files/figure-html/unnamed-chunk-29-1.png" width="672"></p>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>Melbourne Integrative Genomics, The University of Melbourne (<a href="mailto:al.jal.abadi@gmail.com" class="email">al.jal.abadi@gmail.com</a>)<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Al JalalAbadi.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
