<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>scNMTseq challenge analysis using a PLS-based approach • BIRSBIO2020.scNMTseq.PLS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="scNMTseq challenge analysis using a PLS-based approach">
<meta property="og:description" content="BIRSBIO2020.scNMTseq.PLS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BIRSBIO2020.scNMTseq.PLS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/index.html">scNMTseq challenge analysis using a PLS-based approach</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ajabadi/BIRSBIO2020.scNMTseq.PLS">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="index_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>scNMTseq challenge analysis using a PLS-based approach</h1>
                        <h4 class="author">Al JalalAbadi<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</h4>
            
            <h4 class="date">21 Jul 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ajabadi/BIRSBIO2020.scNMTseq.PLS/blob/master/vignettes/index.Rmd"><code>vignettes/index.Rmd</code></a></small>
      <div class="hidden name"><code>index.Rmd</code></div>

    </div>

    
    
<p>To Do</p>
<ul>
<li>Add performance measure</li>
<li>Add the number of features to subtitles in <code>cov_plot</code>
</li>
<li>Ensure python3 can be configured for MOFA2</li>
</ul>
<hr>
<p><strong>Note:</strong> These analyses are (slightly) different from the ones presented in the hackathon in the sense that:</p>
<ul>
<li>The cells used are different (only embryonic cells used)</li>
<li>To enhance reproducibility on different machines, <code><a href="https://rdrr.io/pkg/uwot/man/umap.html">uwot::umap</a></code> was used instead of <code><a href="https://rdrr.io/pkg/umap/man/umap.html">umap::umap</a></code> for dimensionality reduction of RNA data so that python depndencies are reduced</li>
</ul>
<hr>
<p>Load the required packages:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BIRSBIO2020.scNMTseq.PLS</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">MultiAssayExperiment</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">mixOmics</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">knitr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">uwot</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">nipals</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">biomaRt</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">magrittr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">reshape2</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">data.table</span>)</pre></body></html></div>
<!-- ```{r, eval=params$save_output, include=params$save_output} -->
<!-- ## create directories to save the figures and output data: -->
<!-- if (!dir.exists('figures')) { -->
<!--   cat('creating "figures" folder ...\n') -->
<!--   dir.create('figures') -->
<!-- } -->
<!-- if (!dir.exists('savedata')) { -->
<!--    cat('creating "savedata" folder ...\n') -->
<!--   dir.create('savedata') -->
<!-- } -->
<!-- ``` -->
<!-- # Data -->
<!-- Details of the hackathon data and preprocessing steps: https://github.com/BIRSBiointegration/Hackathon/tree/master/scNMT-seq -->
<!-- Load the `MultiAssayExperiment` object: -->
<!-- ```{r, eval=params$on_my_mac, include=FALSE} -->
<!-- cat('loading data from Cloudstor ...\n') -->
<!-- gastru.mae <- readRDS('/Users/alabadi/Projects/analysis/R/multiOmics/scNMTseq/BIRSBIO_scNMTseq/output/scnmtseq_gastrulation_mae_826-cells_orderedFeatures-weights.rds') -->
<!-- ``` -->
<!-- ```{r, eval=params$not_on_my_mac} -->
<!-- cat(sprintf('loading data from %s ...\n', params$data_path)) -->
<!-- gastru.mae <- readRDS(url(params$data_path)) -->
<!-- ``` -->
<!-- An overview of the data object: -->
<!-- ```{r} -->
<!-- gastru.mae -->
<!-- ``` -->
<!-- ## filter extra-embryonic cells -->
<!-- ```{r} -->
<!-- cat(sprintf('Dropping cells from lineage(s) %s ... \n', paste(sQuote(params$drop_lineages), collapse = ', '))) -->
<!-- cat(sprintf('Retaining assays: %s ... \n', paste(sQuote(params$keep_assays), collapse = ', '))) -->
<!-- gastru.mae <- gastru.mae[, !(gastru.mae$lineage10x_2 %in% params$drop_lineages), params$keep_assays] -->
<!-- ``` -->
<!-- ```{r} -->
<!-- gastru.mae -->
<!-- ``` -->
<!-- Breakdown of the cells in each stage: -->
<!-- ```{r} -->
<!-- table(gastru.mae$stage) %>% as.data.frame() %>% set_colnames(c('stage', '# of cells')) %>% kable() -->
<!-- ## create a data.frame from cell meta data  -->
<!-- coldata <- data.frame(colData(gastru.mae)) -->
<!-- ``` -->
<!-- ## Feature detection -->
<!-- Create density plots of the feature detection rate across all cells for all modalities: -->
<!-- ```{r} -->
<!-- # get the methylation assays -->
<!-- met_assays <- grep(names(gastru.mae), pattern = '^met', value = TRUE) -->
<!-- # calculate the feature detection in a data.frame for methylation assays -->
<!-- coverages <- lapply(named_list(met_assays), function(assay_name) { -->
<!--   mat <- assay(gastru.mae, assay_name) -->
<!--   NAs <- rowSums(!is.na(mat))/dim(mat)[2]*100 -->
<!--   data.frame(pct_NAs=NAs) -->
<!-- }) -->
<!-- # create a long data.frame containing the assay name for plot -->
<!-- coverages <- rbindListWithNames(coverages) -->
<!-- coverages$dataset <- factor(coverages$dataset, levels = unique(coverages$dataset), ordered = TRUE) -->
<!-- ``` -->
<!-- ```{r, fig.width=8, fig.asp=0.4, fig.cap="density plots for methylation data show that shorter genomic regions tend to have less feture coverage. Dashed blue line indicates the average across all modalities."} -->
<!-- cov_plot <- ggplot(coverages, aes(x = pct_NAs)) + geom_density(fill = 'lightblue', show.legend = FALSE) +  -->
<!--   geom_vline(aes(xintercept=mean(pct_NAs)), -->
<!--              color="blue", linetype="dashed", size=0.5) + -->
<!--   labs(x = '% of cells detecting the feature') + facet_wrap(.~dataset, nrow = 2) +  -->
<!--   theme_bw() + theme(strip.text.x = element_text(size = 10, face = 'bold')) -->
<!-- cov_plot -->
<!-- ``` -->
<!-- ```{r, eval=params$save_output, include=params$save_output} -->
<!-- ggsave(cov_plot, filename = 'figures/covplots.pdf', width = 8, height = 4) -->
<!-- ``` -->
<!-- # RNA -->
<!-- ## UMAP -->
<!-- ```{r} -->
<!-- RNGversion('4.0.0') -->
<!-- set.seed(params$umap_params['run.seed']) -->
<!-- ## all params use defaults except for 'n_components' and 'min_dist' -->
<!-- cat(sprintf('running umap on RNA data with parameters: %s ...\n ',  -->
<!--             paste(names(params$umap_params[-1]), ':',  -->
<!--                   params$umap_params[-1], collapse = ', '))) -->
<!-- umap_rna_unsup <- uwot::umap( -->
<!--   t(assay(gastru.mae, 'rna')),  -->
<!--   n_neighbors = params$umap_params['n_neighbors'], -->
<!--   n_components = params$umap_params['n_components'], -->
<!--   metric = "euclidean", -->
<!--   n_epochs = NULL, -->
<!--   learning_rate = 1, -->
<!--   scale = FALSE, -->
<!--   init = "spectral", -->
<!--   init_sdev = NULL, -->
<!--   spread = 1, -->
<!--   min_dist = params$umap_params['min_dist'], -->
<!--   set_op_mix_ratio = 1, -->
<!--   local_connectivity = 1) -->
<!-- ``` -->
<!-- ```{r, eval=params$save_output, include=params$save_output} -->
<!-- saveRDS(umap_rna_unsup, file = sprintf('savedata/umap-result-ncomp-%s.rds', ncomp)) -->
<!-- ``` -->
<!-- Create a data.frame of embeddings: -->
<!-- ```{r} -->
<!-- df <- data.frame(umap_rna_unsup) -->
<!-- colnames(df) <- paste0('UMAP_', seq_along(df)) -->
<!-- df <- cbind(df, coldata) -->
<!-- ``` -->
<!-- Create colour palettes for stages and lineages: -->
<!-- ```{r} -->
<!-- col_pallete <-c("#440154FF", "#31688EFF", "#35B779FF", "#FDE725FF") -->
<!-- names(col_pallete) <- c("E4.5", "E5.5", "E6.5", "E7.5") -->
<!-- ``` -->
<!-- ```{r} -->
<!-- ## helper function to create both plots coloured by stage -->
<!-- plot_umap_by_stage <- function(df, dims = c(1,2)) { -->
<!--   axes <- paste0('UMAP_', dims) -->
<!--   ggplot(df, aes_string(axes[1], axes[2])) + geom_point(aes(col=stage)) +  -->
<!--     theme_classic()+ scale_color_manual(values = col_pallete) -->
<!-- } -->
<!-- ``` -->
<!-- First two components highlight the distinct transcriptional state of the early-stage cells compared to the late-stage cells: -->
<!-- ```{r, eval=FALSE, fig.asp=0.6} -->
<!-- ## dimensions 1 to 3 capture early- vs late-stage variation -->
<!-- plot_umap_by_stage(df = df, dims = c(1,2)) -->
<!-- plot_umap_by_stage(df = df, dims = c(2,3)) -->
<!-- ``` -->
<!-- The variation driven by different stages is more apparent in the following dimensions: -->
<!-- ```{r, fig.asp=0.6, fig.cap="UMAP plot shows distinct cell populations corresponding to early-stage, mid-stage, and late-stage cells as well as the within-stage variation which could correspond to different lineages."} -->
<!-- plot_umap_by_stage(df = df, dims = c(3,4)) -->
<!-- ``` -->
<!-- # Integration -->
<!-- Get all the gene names from all modalities: -->
<!-- ```{r} -->
<!-- all_features <- sapply(experiments(gastru.mae), rownames) %>% unlist() -->
<!-- all_genes <- grep(pattern = '^ENSMUSG', x = all_features, value = TRUE) -->
<!-- all_genes <- unique(all_genes) -->
<!-- length(all_genes) -->
<!-- ``` -->
<!-- Get the Ensembl gene name for these gene ids: -->
<!-- ```{r} -->
<!-- ensembl <- useMart("ensembl", dataset="mmusculus_gene_ensembl") -->
<!-- mouse_gene_ids  <- all_genes -->
<!-- symbols <- getBM(attributes=c('ensembl_gene_id', -->
<!--                           'external_gene_name'), -->
<!--              filters = "ensembl_gene_id", -->
<!--              values = all_genes, -->
<!--              mart = ensembl) -->
<!-- symbols <- data.frame(symbols, row.names = 1) -->
<!-- ``` -->
<!-- ```{r, eval=params$save_output, echo=FALSE} -->
<!-- save(symbols, file = 'savedata/all-gene-symbols.RData') -->
<!-- # load('savedata/all-gene-symbols.RData') -->
<!-- ``` -->
<!-- ```{r} -->
<!-- lapply(experiments(gastru.mae), dim) -->
<!-- ``` -->
<!-- ```{r, eval=FALSE} -->
<!-- mmspls <- multimodal_analysis_wrapper(mae = gastru.mae, study_assays = NULL, ncomp = 2, scale = FALSE, design = 'null', lineages = NULL, stages = NULL, DA = NULL, keepX = NULL, save = FALSE) -->
<!-- ``` -->
<!-- ```{r, eval=params$save_output, echo=FALSE} -->
<!-- saveRDS(mmspls, file = 'savedata/MultiModalSparsePLS-All.rds') -->
<!-- # mmspls <- readRDS('savedata/MultiModalSparsePLS-All.rds') -->
<!-- ``` -->
<!-- ```{r} -->
<!-- plotIndiv_wrapper <- function(pls_obj, col = 'stage', legend.title = 'Stage', comp = c(1,2), ...) { -->
<!--   plotIndiv(pls_obj, pch = 16, comp = comp,group = droplevels(factor(coldata[rownames(pls_obj$X[[1]]),][,col])), legend = TRUE, legend.title = legend.title, size.subtitle = 10, cex=0.8, ...) -->
<!-- } -->
<!-- ``` -->
<!-- ## score plots -->
<!-- Colored by stage: -->
<!-- ```{r} -->
<!-- plotIndiv_wrapper(pls_obj = mmspls,col = 'stage', legend.title = 'Stage', comp = c(1,2), subtitle = names(mmspls$X), col.per.group = col_pallete) -->
<!-- ``` -->
<!-- ```{r, eval=params$save_output, echo=FALSE} -->
<!-- ggsave(filename = 'figures/plotIndiv-MultiModalSparsePLS-stage.pdf', width = 8, height = 6) -->
<!-- ``` -->
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>Melbourne Integrative Genomics, The University of Melbourne (<a href="mailto:al.jal.abadi@gmail.com" class="email">al.jal.abadi@gmail.com</a>)<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Al JalalAbadi.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
