<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>scNMTseq challenge analysis using a PLS-based approach • BIRSBIO2020.scNMTseq.PLS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="scNMTseq challenge analysis using a PLS-based approach">
<meta property="og:description" content="BIRSBIO2020.scNMTseq.PLS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BIRSBIO2020.scNMTseq.PLS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/index.html">scNMTseq challenge analysis using a PLS-based approach</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ajabadi/BIRSBIO2020.scNMTseq.PLS">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="index_files/header-attrs-2.8/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>scNMTseq challenge analysis using a PLS-based approach</h1>
                        <h4 class="author">Al JalalAbadi<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</h4>
            
            <h4 class="date">19 May 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ajabadi/BIRSBIO2020.scNMTseq.PLS/blob/master/vignettes/index.Rmd"><code>vignettes/index.Rmd</code></a></small>
      <div class="hidden name"><code>index.Rmd</code></div>

    </div>

    
    
<hr>
<p><strong>Note:</strong> These analyses are different from the ones presented in the hackathon due to updated preprocessing if data in order to harmonize the analyses for publication. For original analyses see <a href="https://github.com/ajabadi/scNMT_seq_gastrulation" class="uri">https://github.com/ajabadi/scNMT_seq_gastrulation</a>.</p>
<hr>
<p>Load the required packages:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ajabadi.github.io/BIRSBIO2020.scNMTseq.PLS/">BIRSBIO2020.scNMTseq.PLS</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://waldronlab.io/MultiAssayExperiment/">MultiAssayExperiment</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://www.mixOmics.org">mixOmics</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org">magrittr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hadley/reshape">reshape2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jlmelville/uwot">uwot</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">impute</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://kwstat.github.io/nipals/">nipals</a></span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">nipals_maxiter</span> <span class="op">&lt;-</span> <span class="va">params</span><span class="op">$</span><span class="va">nipals_params</span><span class="op">[</span><span class="st">'maxiter'</span><span class="op">]</span>
<span class="va">nipals_ncomp</span> <span class="op">&lt;-</span> <span class="va">params</span><span class="op">$</span><span class="va">nipals_params</span><span class="op">[</span><span class="st">'ncomp'</span><span class="op">]</span>
<span class="va">nipals_nhvr</span> <span class="op">&lt;-</span> <span class="va">params</span><span class="op">$</span><span class="va">nipals_params</span><span class="op">[</span><span class="st">'nhvr'</span><span class="op">]</span></code></pre></div>
<div id="data" class="section level1">
<h1 class="hasAnchor">
<a href="#data" class="anchor"></a>Data</h1>
<p>Details of the hackathon data and preprocessing steps: <a href="https://github.com/BIRSBiointegration/Hackathon/tree/master/scNMT-seq" class="uri">https://github.com/BIRSBiointegration/Hackathon/tree/master/scNMT-seq</a></p>
<p>Load the <code>MultiAssayExperiment</code> object from Cloudstor:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">'loading data from Cloudstor ...\n'</span><span class="op">)</span>
<span class="co">#&gt; loading data from Cloudstor ...</span>
<span class="va">gastru.mae_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/connections.html">url</a></span><span class="op">(</span><span class="st">'https://cloudstor.aarnet.edu.au/plus/s/jsW7nh4YwThw8Q5/download'</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gastru.mae</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="va">gastru.mae_path</span><span class="op">)</span></code></pre></div>
<!-- ```{r, eval=!params$on_my_mac} -->
<!-- cat(sprintf('loading the following assays using SingleCellMultiModal package: %s\n', paste(params$keep_assays, collapse = ', '))) -->
<!-- gastru.mae <- scNMT(dataType = 'mouse_gastrulation', modes = params$keep_assays, dry.run = FALSE, verbose = FALSE) -->
<!-- ``` -->
<div id="filter-mae-object" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-mae-object" class="anchor"></a>Filter MAE object</h2>
<p>Keep cells which pass QC metrics and exclude putative extra-embryonic cells as they are starkly different in methylation patterns and will drive most of variation in the relatively small population of matching cells, not allowing to fully explore cell to cell variation intended for the purpose of this hackathon:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## subset RNA expression and DNA methylation modalities</span>
<span class="va">keep_assays</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"rna|met"</span>,<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu">assays</span><span class="op">(</span><span class="va">gastru.mae</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Warning: Dropping internal assays in 'rna'; taking first one</span>
<span class="va">gastru.mae</span> <span class="op">&lt;-</span> <span class="va">gastru.mae</span><span class="op">[</span>,,<span class="va">keep_assays</span><span class="op">]</span>
<span class="co">#&gt; Warning: 'experiments' dropped; see 'metadata'</span>
<span class="co">#&gt; harmonizing input:</span>
<span class="co">#&gt;   removing 4403 sampleMap rows not in names(experiments)</span>
<span class="co">## remove putative extraembryonic cells</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span><span class="op">(</span><span class="st">'dropping lineages %s, plus the unassigned lineages.\n'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">drop_lineages</span>, collapse <span class="op">=</span> <span class="st">', '</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; dropping lineages Primitive_endoderm, Visceral_endoderm, ExE_ectoderm, plus the unassigned lineages.</span>
<span class="va">filter_cells</span> <span class="op">&lt;-</span> <span class="va">gastru.mae</span><span class="op">$</span><span class="va">lineage</span> <span class="op">%in%</span> <span class="va">params</span><span class="op">$</span><span class="va">drop_lineages</span>
<span class="va">filter_cells</span> <span class="op">&lt;-</span> <span class="va">filter_cells</span> <span class="op">|</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">gastru.mae</span><span class="op">$</span><span class="va">lineage</span><span class="op">)</span>
<span class="co">## outlier cell to be filtered</span>
<span class="va">filter_cells</span> <span class="op">&lt;-</span> <span class="va">filter_cells</span> <span class="op">|</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">gastru.mae</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="st">'E6.75_Plate2_H10'</span>
<span class="va">gastru.mae</span> <span class="op">&lt;-</span> <span class="va">gastru.mae</span><span class="op">[</span>,<span class="op">!</span><span class="va">filter_cells</span>,<span class="op">]</span>
<span class="co">## keep cells which pass RNA QC</span>
<span class="va">gastru.mae</span> <span class="op">&lt;-</span> <span class="va">gastru.mae</span><span class="op">[</span>,<span class="va">gastru.mae</span><span class="op">$</span><span class="va">pass_rnaQC</span><span class="op">==</span><span class="cn">TRUE</span>,<span class="op">]</span>
<span class="co">## Keep full rna SCE for UMAP</span>
<span class="va">rna.sce</span> <span class="op">&lt;-</span> <span class="va">gastru.mae</span><span class="op">@</span><span class="va">ExperimentList</span><span class="op">$</span><span class="va">rna</span>
<span class="co">## keep cells that also pass QC for DNA methylation</span>
<span class="va">gastru.mae</span> <span class="op">&lt;-</span> <span class="va">gastru.mae</span><span class="op">[</span>,<span class="va">gastru.mae</span><span class="op">$</span><span class="va">pass_metQC</span><span class="op">==</span><span class="cn">TRUE</span>,<span class="op">]</span>
<span class="co">## rna SCE for cells passing met and rna QC</span>
<span class="va">rna.sce.matching</span> <span class="op">&lt;-</span> <span class="va">gastru.mae</span><span class="op">@</span><span class="va">ExperimentList</span><span class="op">$</span><span class="va">rna</span></code></pre></div>
<p>Replace the rna SCE with normalised counts in MAE object as the integration wrapper requires matrices in assays:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## rna is an SCE object</span>
<span class="va">gastru.mae</span><span class="op">@</span><span class="va">ExperimentList</span>
<span class="co">#&gt; ExperimentList class object of length 5:</span>
<span class="co">#&gt;  [1] rna: SingleCellExperiment with 18345 rows and 795 columns</span>
<span class="co">#&gt;  [2] met_genebody: matrix with 17559 rows and 795 columns</span>
<span class="co">#&gt;  [3] met_promoter: matrix with 17179 rows and 795 columns</span>
<span class="co">#&gt;  [4] met_cgi: matrix with 14080 rows and 795 columns</span>
<span class="co">#&gt;  [5] met_DHS: matrix with 6673 rows and 795 columns</span>
<span class="co">## replace SCE with logcounts</span>
<span class="va">gastru.mae</span><span class="op">@</span><span class="va">ExperimentList</span><span class="op">$</span><span class="va">rna</span> <span class="op">&lt;-</span> <span class="fu">logcounts</span><span class="op">(</span><span class="va">gastru.mae</span><span class="op">@</span><span class="va">ExperimentList</span><span class="op">$</span><span class="va">rna</span><span class="op">)</span>
<span class="co">## rna updated to matrix of logcounts</span>
<span class="va">gastru.mae</span><span class="op">@</span><span class="va">ExperimentList</span>
<span class="co">#&gt; ExperimentList class object of length 5:</span>
<span class="co">#&gt;  [1] rna: matrix with 18345 rows and 795 columns</span>
<span class="co">#&gt;  [2] met_genebody: matrix with 17559 rows and 795 columns</span>
<span class="co">#&gt;  [3] met_promoter: matrix with 17179 rows and 795 columns</span>
<span class="co">#&gt;  [4] met_cgi: matrix with 14080 rows and 795 columns</span>
<span class="co">#&gt;  [5] met_DHS: matrix with 6673 rows and 795 columns</span></code></pre></div>
<p>Breakdown of the number of cells in each stage:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">gastru.mae</span><span class="op">$</span><span class="va">stage</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://magrittr.tidyverse.org/reference/aliases.html">set_rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'stage'</span>, <span class="st">'# of cells'</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">kable</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<table class="table"><tbody>
<tr class="odd">
<td align="left">stage</td>
<td align="left">E4.5</td>
<td align="left">E5.5</td>
<td align="left">E6.5</td>
<td align="left">E7.5</td>
</tr>
<tr class="even">
<td align="left"># of cells</td>
<td align="left">62</td>
<td align="left">86</td>
<td align="left">274</td>
<td align="left">373</td>
</tr>
</tbody></table>
</div>
<div id="feature-detection" class="section level2">
<h2 class="hasAnchor">
<a href="#feature-detection" class="anchor"></a>Feature detection</h2>
<p>Create density plots of the feature detection rate across all cells for all modalities:</p>
<pre><code>#&gt; Warning: 'experiments' dropped; see 'metadata'</code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cov_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">coverages</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">pct_NAs</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">'lightblue'</span>, show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>xintercept<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">pct_NAs</span><span class="op">)</span><span class="op">)</span>,
             color<span class="op">=</span><span class="st">"blue"</span>, linetype<span class="op">=</span><span class="st">"dashed"</span>, size<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">'% of cells detecting the feature'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="va">.</span><span class="op">~</span><span class="va">dataset</span>, nrow <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>strip.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span>, face <span class="op">=</span> <span class="st">'bold'</span>, color <span class="op">=</span> <span class="st">'purple'</span><span class="op">)</span><span class="op">)</span>
  
<span class="va">cov_plot</span></code></pre></div>
<p><img src="index_files/figure-html/unnamed-chunk-11-1.png" width="768"></p>
<p>Density plots for methylation data show that shorter genomic regions tend to have less feature coverage. Dashed blue line indicates the average across all modalities.</p>
</div>
</div>
<div id="rna" class="section level1">
<h1 class="hasAnchor">
<a href="#rna" class="anchor"></a>RNA</h1>
<p>Run PCA and then UMAP using PCs:</p>
<div id="cells-passing-rna-qc-n-2146" class="section level2">
<h2 class="hasAnchor">
<a href="#cells-passing-rna-qc-n-2146" class="anchor"></a>Cells passing rna QC (n = 2146)</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">npcs</span> <span class="op">&lt;-</span> <span class="fl">15</span>
<span class="co">## PCA first: retrieve npcs PCs</span>
<span class="va">rna.sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runPCA.html">runPCA</a></span><span class="op">(</span><span class="va">rna.sce</span>,  ncomponents <span class="op">=</span> <span class="va">npcs</span>, name <span class="op">=</span> <span class="st">'PCA'</span><span class="op">)</span>

<span class="co">## UMAP parameters used:</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span><span class="op">(</span><span class="st">'Running UMAP with parameters %s\n'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">umap_params</span><span class="op">)</span>, <span class="st">':'</span>,<span class="va">params</span><span class="op">$</span><span class="va">umap_params</span>, collapse <span class="op">=</span> <span class="st">', '</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Running UMAP with parameters run.seed : 42, n_neighbors : 15, n_components : 2, min_dist : 0.55</span>

<span class="co">## run UMAP</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">umap_params</span><span class="op">[</span><span class="st">'run.seed'</span><span class="op">]</span><span class="op">)</span>
<span class="va">rna.sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">rna.sce</span>, dimred<span class="op">=</span><span class="st">"PCA"</span>, 
                   ncomponents <span class="op">=</span> <span class="va">params</span><span class="op">$</span><span class="va">umap_params</span><span class="op">[</span><span class="st">'n_components'</span><span class="op">]</span>, 
                   n_neighbors <span class="op">=</span> <span class="va">params</span><span class="op">$</span><span class="va">umap_params</span><span class="op">[</span><span class="st">'n_neighbors'</span><span class="op">]</span>, 
                   min_dist <span class="op">=</span> <span class="va">params</span><span class="op">$</span><span class="va">umap_params</span><span class="op">[</span><span class="st">'min_dist'</span><span class="op">]</span><span class="op">)</span></code></pre></div>
</div>
<div id="cells-passing-both-rna-and-methylation-qc-n-795" class="section level2">
<h2 class="hasAnchor">
<a href="#cells-passing-both-rna-and-methylation-qc-n-795" class="anchor"></a>Cells passing both rna and methylation QC (n = 795)</h2>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">decomp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html">modelGeneVar</a></span><span class="op">(</span><span class="va">rna.sce.matching</span><span class="op">)</span>
<span class="co">## filter by mean expression and significance of biological variation signal</span>
<span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">decomp</span><span class="op">)</span><span class="op">[</span><span class="va">decomp</span><span class="op">$</span><span class="va">p.value</span><span class="op">&lt;</span><span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">decomp</span><span class="op">$</span><span class="va">mean</span> <span class="op">&gt;</span> <span class="fl">0.01</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">hvgs</span><span class="op">)</span>
<span class="co">#&gt; [1] 680</span>

<span class="va">npcs</span> <span class="op">&lt;-</span> <span class="fl">15</span>
<span class="co">## PCA first: retrieve npcs PCs</span>
<span class="va">rna.sce.matching</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runPCA.html">runPCA</a></span><span class="op">(</span><span class="va">rna.sce.matching</span>, ncomponents <span class="op">=</span> <span class="va">npcs</span>, subset_row<span class="op">=</span><span class="va">hvgs</span>, name <span class="op">=</span> <span class="st">'PCA'</span><span class="op">)</span>

<span class="co">## UMAP parameters used:</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span><span class="op">(</span><span class="st">'Running UMAP with parameters %s\n'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">umap_params</span><span class="op">)</span>, <span class="st">':'</span>,<span class="va">params</span><span class="op">$</span><span class="va">umap_params</span>, collapse <span class="op">=</span> <span class="st">', '</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Running UMAP with parameters run.seed : 42, n_neighbors : 15, n_components : 2, min_dist : 0.55</span>

<span class="co">## run UMAP</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">umap_params</span><span class="op">[</span><span class="st">'run.seed'</span><span class="op">]</span><span class="op">)</span>
<span class="va">rna.sce.matching</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">rna.sce.matching</span>, dimred<span class="op">=</span><span class="st">"PCA"</span>, 
                   ncomponents <span class="op">=</span> <span class="va">params</span><span class="op">$</span><span class="va">umap_params</span><span class="op">[</span><span class="st">'n_components'</span><span class="op">]</span>, 
                   n_neighbors <span class="op">=</span> <span class="va">params</span><span class="op">$</span><span class="va">umap_params</span><span class="op">[</span><span class="st">'n_neighbors'</span><span class="op">]</span>, 
                   min_dist <span class="op">=</span> <span class="va">params</span><span class="op">$</span><span class="va">umap_params</span><span class="op">[</span><span class="st">'min_dist'</span><span class="op">]</span><span class="op">)</span></code></pre></div>
</div>
<div id="umap" class="section level2">
<h2 class="hasAnchor">
<a href="#umap" class="anchor"></a>UMAP</h2>
<p>Plot the first two UMAP components:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create colour palettes for stages:</span>
<span class="va">stages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"E4.5"</span>, <span class="st">"E5.5"</span>, <span class="st">"E6.5"</span>, <span class="st">"E7.5"</span><span class="op">)</span>
<span class="co"># col_palette &lt;- dput(viridisLite::viridis(n = length(stages)))</span>
<span class="va">col_palette</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#440154FF"</span>, <span class="st">"#31688EFF"</span>, <span class="st">"#35B779FF"</span>, <span class="st">"#FDE725FF"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">col_palette</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">stages</span></code></pre></div>
<div id="all-cells" class="section level3">
<h3 class="hasAnchor">
<a href="#all-cells" class="anchor"></a>all cells</h3>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gg_rna_all</span> <span class="op">&lt;-</span> <span class="fu">plot_reducedDim</span><span class="op">(</span>sce <span class="op">=</span> <span class="va">rna.sce</span>, reducedDim <span class="op">=</span> <span class="st">'UMAP'</span>, dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, col_palette <span class="op">=</span> <span class="va">col_palette</span><span class="op">)</span>
<span class="va">gg_rna_all</span></code></pre></div>
<p><img src="index_files/figure-html/unnamed-chunk-14-1.png" width="480" style="display: block; margin: auto;"></p>
<p>UMAP plot of all cells passing rna QC controls separates the cell populations from early-stage, mid-stage, and late-stage cells. This structure is also apparent in the matching cells’ local embeddings but less strongly:</p>
</div>
<div id="matching-cells" class="section level3">
<h3 class="hasAnchor">
<a href="#matching-cells" class="anchor"></a>matching cells</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gg_rna_matching</span> <span class="op">&lt;-</span> <span class="fu">plot_reducedDim</span><span class="op">(</span>sce <span class="op">=</span> <span class="va">rna.sce.matching</span>, reducedDim <span class="op">=</span> <span class="st">'UMAP'</span>, dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, col_palette <span class="op">=</span> <span class="va">col_palette</span><span class="op">)</span>
<span class="va">gg_rna_matching</span></code></pre></div>
<p><img src="index_files/figure-html/unnamed-chunk-15-1.png" width="480" style="display: block; margin: auto;"></p>
<!-- We now visualise the putative lineages to see whether the population structure is in line with the mapped lineages using data from all cells: -->
<!-- ```{r, fig.asp=0.7, fig.width=5, fig.align='center'} -->
<!-- ## colour by lineage -->
<!-- rna.sce.matching$lineage <- gastru.mae$lineage -->
<!-- rna.sce.matching$lineage[rna.sce.matching$lineage == 'NOIDEA'] <- NA -->
<!-- plot_reducedDim(sce = rna.sce.matching, reducedDim = 'UMAP', dims = c(1,2), col_palette = NULL, col = 'lineage') -->
<!-- ``` -->
<!-- We can see that the variation in the late-stage cells is driven by disparate germ layer cells. -->
</div>
</div>
</div>
<div id="dna" class="section level1">
<h1 class="hasAnchor">
<a href="#dna" class="anchor"></a>DNA</h1>
<div id="empca" class="section level2">
<h2 class="hasAnchor">
<a href="#empca" class="anchor"></a>EMPCA</h2>
<p>We perform PCA using an Expectation Maximization algorithm (NIPALS) for all methylation measurements using highly variable regions:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">met_assays</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">gastru.mae</span><span class="op">@</span><span class="va">ExperimentList</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>
<span class="va">nipals_comps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">M</span> <span class="kw">in</span> <span class="va">met_assays</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">gastru.mae</span><span class="op">@</span><span class="va">ExperimentList</span><span class="op">[[</span><span class="va">M</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
  <span class="va">nhvr</span> <span class="op">&lt;-</span> <span class="va">nipals_nhvr</span>
  <span class="va">d</span> <span class="op">&lt;-</span> <span class="va">d</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="op">-</span><span class="fu">colVars</span><span class="op">(</span><span class="va">d</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">&lt;=</span><span class="va">nhvr</span><span class="op">]</span>
  <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">'Performing nipals on:'</span>, <span class="va">M</span>, <span class="st">'\n'</span><span class="op">)</span>
  <span class="va">nipals_comps</span><span class="op">[[</span><span class="va">M</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/r/base/warning.html">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span>
      <span class="fu">nipals</span><span class="fu">::</span><span class="fu"><a href="http://kwstat.github.io/nipals/reference/nipals.html">nipals</a></span><span class="op">(</span><span class="va">d</span>, 
                     ncomp <span class="op">=</span> <span class="va">nipals_ncomp</span>, 
                     center <span class="op">=</span> <span class="cn">TRUE</span>,
                     scale <span class="op">=</span> <span class="cn">FALSE</span>,
                     <span class="va">nipals_maxiter</span><span class="op">)</span>
    <span class="op">}</span><span class="op">)</span>
    
<span class="op">}</span>
<span class="co">#&gt; Performing nipals on: met_genebody </span>
<span class="co">#&gt; Performing nipals on: met_promoter </span>
<span class="co">#&gt; Performing nipals on: met_cgi </span>
<span class="co">#&gt; Performing nipals on: met_DHS</span>
<span class="co"># cat('\nRunning nipals on', parallel::detectCores(), ' cpus...\n')</span>
<span class="co"># nipals_comps &lt;- mclapply(named_list(met_assays), FUN = function(M){</span>
<span class="co">#   nipals::nipals(t(gastru.mae@ExperimentList[[M]]), </span>
<span class="co">#                  ncomp = 2, center = TRUE, scale = FALSE, </span>
<span class="co">#                  maxiter = ifelse(params$mini_run, 10, 500))</span>
<span class="co"># }, mc.cores = parallel::detectCores())</span></code></pre></div>
</div>
<div id="score-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#score-plots" class="anchor"></a>Score plots</h2>
<p>Plot PCs:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">plot_nipals</span><span class="op">(</span><span class="va">nipals_comps</span>, cell_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">gastru.mae</span><span class="op">@</span><span class="va">ExperimentList</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, 
            stage <span class="op">=</span> <span class="va">gastru.mae</span><span class="op">$</span><span class="va">stage</span>, red_dim <span class="op">=</span> <span class="st">'PC'</span>, col_palette <span class="op">=</span> <span class="va">col_palette</span>, 
            dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, facet_ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<p><img src="index_files/figure-html/plot%20nipals%20PCs-1.png" width="768"></p>
<p>The first two components from Expectation Maximization PCA. Whereas these PCs mainly distinguish early-stage from late-stage cells in all methylation views, an observation in agreement with the transcriptome UMAP projection, the stages are better resolved in some views (e.g. met_cgi) than others.</p>
<!-- Inspect the outlier cell in promoter view: -->
<!-- ```{r plot nipals PCs with outlier, fig.width=8, fig.height=6} -->
<!-- outlier <- which.max(nipals_comps$met_promoter$scores[,1]) -->
<!-- plot_nipals(nipals_comps, cell_order = colnames(gastru.mae@ExperimentList[[1]]),  -->
<!--             stage = gastru.mae$stage, red_dim = 'PC', col_palette = col_palette,  -->
<!--             dims = c(1, 2), facet_ncol = 2, show.cell = 'E6.75_Plate2_H10') -->
<!-- ``` -->
<!-- The cell is also an outlier in the genebody view. Let's look at the the amount of missing values the cell measurements contain: -->
<!-- ```{r} -->
<!-- ## missing values -->
<!-- sapply(gastru.mae@ExperimentList[2:5], function(E){ -->
<!--   sum(is.na(E[,'E6.75_Plate2_H10']))/length(E[,'E6.75_Plate2_H10']) -->
<!-- }) -->
<!-- ``` -->
<!-- Interestingly, this cell has least amount of missing values in views where it seems to be an outlier. Let's see what happens if we filter it. -->
<!-- ## Filter outlier -->
<!-- ```{r} -->
<!-- el <- gastru.mae@ExperimentList -->
<!-- ind_outlier <- which(colnames(el[[1]]) == 'E6.75_Plate2_H10') -->
<!-- el <- lapply(el, function(E) { -->
<!--   E[,-ind_outlier] -->
<!-- }) -->
<!-- met_assays <- names(el)[-1] -->
<!-- nipals_comps_filt <- list() -->
<!-- for (M in met_assays) { -->
<!--   d <- t(el[[M]]) -->
<!--   nhvr <- nipals_nhvr -->
<!--   d <- d[,order(-colVars(d, na.rm = TRUE))<=nhvr] -->
<!--   cat('Performing nipals on:', M, '\n') -->
<!--   nipals_comps_filt[[M]] <-  -->
<!--     suppressWarnings({ -->
<!--       nipals::nipals(d,  -->
<!--                      ncomp = nipals_ncomp,  -->
<!--                      center = TRUE, -->
<!--                      scale = FALSE, -->
<!--                      nipals_maxiter) -->
<!--     }) -->
<!-- } -->
<!-- ``` -->
<!-- ```{r plot nipals PCs without outlier, fig.width=8, fig.height=6} -->
<!-- plot_nipals(nipals_comps_filt, cell_order = colnames(el[[1]]),  -->
<!--             stage = gastru.mae$stage[-ind_outlier], red_dim = 'PC', col_palette = col_palette,  -->
<!--             dims = c(1, 2), facet_ncol = 2) -->
<!-- ``` -->
<!-- There seems to be an improvement in the ability of components to separate the stages. We remove this cell in downstream analyses: -->
<!-- ```{r} -->
<!-- gastru.mae <- gastru.mae[,-ind_outlier,] -->
<!-- rna.sce.matching <- rna.sce.matching[,-ind_outlier] -->
<!-- ``` -->
</div>
</div>
<div id="integration" class="section level1">
<h1 class="hasAnchor">
<a href="#integration" class="anchor"></a>Integration</h1>
<!-- ## gene id to gene symbol conversion data.frame -->
<!-- Here we create a reference data.frame to convert Ensembl gene ids to gene symbols for GSEA. -->
<!-- Get all the gene names from all modalities: -->
<!-- ```{r} -->
<!-- all_features <- sapply(experiments(gastru.mae), rownames) %>% unlist() -->
<!-- all_genes <- grep(pattern = '^ENSMUSG', x = all_features, value = TRUE) -->
<!-- all_genes <- unique(all_genes) -->
<!-- length(all_genes) -->
<!-- ``` -->
<!-- Get the Ensembl gene name for these gene ids: -->
<!-- ```{r} -->
<!-- ensembl <- useMart("ensembl", dataset="mmusculus_gene_ensembl") -->
<!-- mouse_gene_ids  <- all_genes -->
<!-- symbols <- getBM(attributes=c('ensembl_gene_id', -->
<!--                               'external_gene_name'), -->
<!--                  filters = "ensembl_gene_id", -->
<!--                  values = all_genes, -->
<!--                  mart = ensembl) -->
<!-- symbols <- data.frame(symbols, row.names = 1) -->
<!-- ``` -->
<!-- ```{r, eval=params$save_output, echo=FALSE} -->
<!-- save(symbols, file = 'savedata/all-gene-symbols.RData') -->
<!-- # load('savedata/all-gene-symbols.RData') -->
<!-- ``` -->
<div id="pls" class="section level2">
<h2 class="hasAnchor">
<a href="#pls" class="anchor"></a>PLS</h2>
</div>
<div id="integration-of-all-modalities-with-feature-selection-" class="section level2">
<h2 class="hasAnchor">
<a href="#integration-of-all-modalities-with-feature-selection-" class="anchor"></a>Integration of all modalities with feature selection.</h2>
<p>Here we select for 50 features on each component in <code>multimodal_sPLS_wrapper</code>:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## number of components</span>
<span class="va">ncomp</span> <span class="op">&lt;-</span> <span class="va">params</span><span class="op">$</span><span class="va">pls_ncomp</span>
<span class="co">## feature scaling</span>
<span class="va">scale</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span><span class="op">(</span><span class="st">'Running PLS with %s components performing variable selection\n'</span>, <span class="va">ncomp</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Running PLS with 4 components performing variable selection</span>
<span class="va">st</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span><span class="op">(</span><span class="op">{</span>
  <span class="va">mmspls</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="../reference/multimodal_sPLS_wrapper.html">multimodal_sPLS_wrapper</a></span><span class="op">(</span>mae <span class="op">=</span> <span class="va">gastru.mae</span>, study_assays <span class="op">=</span> <span class="cn">NULL</span>,
                            ncomp <span class="op">=</span> <span class="va">ncomp</span>, scale <span class="op">=</span> <span class="va">scale</span>, design <span class="op">=</span> <span class="st">'null'</span>, lineages <span class="op">=</span> <span class="cn">NULL</span>,
                            stages <span class="op">=</span> <span class="cn">NULL</span>, DA <span class="op">=</span> <span class="cn">NULL</span>, keepX <span class="op">=</span> <span class="cn">NULL</span>, save <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span><span class="op">[</span><span class="st">'elapsed'</span><span class="op">]</span>
<span class="co">#&gt; Making unique feature names by adding assay name to feature name for: met_genebody</span>
<span class="co">#&gt; Making unique feature names by adding assay name to feature name for: met_promoter</span>
<span class="va">mmspls</span><span class="op">$</span><span class="va">stage</span> <span class="op">&lt;-</span> <span class="va">gastru.mae</span><span class="op">$</span><span class="va">stage</span>
<span class="va">mmspls</span><span class="op">$</span><span class="va">lineage</span><span class="op">&lt;-</span> <span class="va">gastru.mae</span><span class="op">$</span><span class="va">lineage</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">'\nPLS run finished. Runtime: '</span>, <span class="fu">format_runtime</span><span class="op">(</span><span class="va">st</span><span class="op">)</span>, <span class="st">'\n'</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; PLS run finished. Runtime:  00:02:21</span></code></pre></div>
</div>
<div id="score-plots-1" class="section level2">
<h2 class="hasAnchor">
<a href="#score-plots-1" class="anchor"></a>score plots</h2>
<p>Plot PLS components for cells across all views colored by embryonic stages:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pls_no.impute</span> <span class="op">&lt;-</span> <span class="fu">plot_pls</span><span class="op">(</span>pls_obj <span class="op">=</span> <span class="va">mmspls</span>, stage <span class="op">=</span> <span class="va">mmspls</span><span class="op">$</span><span class="va">stage</span>, col_palette <span class="op">=</span> <span class="va">col_palette</span><span class="op">)</span>
<span class="va">pls_no.impute</span></code></pre></div>
<p><img src="index_files/figure-html/unnamed-chunk-19-1.png" width="768"></p>
<p>The latent components which seek to maximise the sum of concordant variation between rna and other modalities separate early-stage cells from late-stage cells. It indicates coordinated changes across the methylomes associated with the distinct transcriptional states of cells in the regions of study during gastrulation.</p>
<!-- Colour late-stage cells by putative lineages: -->
<!-- ```{r, fig.width=8, fig.height=5} -->
<!-- keep_stages <- c('E6.5', 'E7.5') -->
<!-- latestage_cells <- mmspls$stage %in% keep_stages -->
<!-- plot_pls(mmspls, mmspls$lineage,  -->
<!--          ind.keep = latestage_cells, -->
<!--          col_palette = gg_color_hue(length(unique(mmspls$lineage))),  -->
<!--          legend.title = 'Lineage') -->
<!-- ``` -->
<!-- We can see the variation within the late-stage cells driven by disparate cell subpopulations in the transcriptome view. -->
<!-- ## UMAP using 4 PLS components -->
<!-- Add PLS embeddings to SCE so UMAP is applied using `scater::runUMAP`: -->
<!-- ```{r} -->
<!-- for (block in names(mmspls$variates)) { -->
<!--   variates <- mmspls$variates[block] -->
<!--   dimred.name <- paste0('PLS_', block) -->
<!--   umap.name <- paste0('UMAP_PLS_', block) -->
<!--   reducedDim(rna.sce.matching, dimred.name) <- mmspls$variates[[block]] -->
<!--   set.seed(params$umap_params['run.seed']) -->
<!--   rna.sce.matching <- runUMAP(rna.sce.matching, dimred=dimred.name,  -->
<!--                               ncomponents = params$umap_params['n_components'],  -->
<!--                               n_neighbors = params$umap_params['n_neighbors'],  -->
<!--                               min_dist = params$umap_params['min_dist'],  -->
<!--                               name = umap.name) -->
<!-- } -->
<!-- ``` -->
<!-- Create a long data.frame of all variates for facet plotting:  -->
<!-- ```{r} -->
<!-- all_pls_umap <- list() -->
<!-- for (block in names(mmspls$variates)) { -->
<!--   umap.name <- paste0('UMAP_PLS_', block) -->
<!--   umap_df <- as.data.frame(reducedDim(rna.sce.matching, umap.name)) -->
<!--   colnames(umap_df) <- paste0('UMAP_', seq_along(umap_df)) -->
<!--   umap_df$stage <- gastru.mae$stage -->
<!--   umap_df$lineage <- gastru.mae$lineage -->
<!--   all_pls_umap[[block]] <- umap_df -->
<!-- } -->
<!-- all_pls_umap <- rbindListWithNames(all_pls_umap, new_col = 'Modality') -->
<!-- ``` -->
<!-- Plot UMAP: -->
<!-- ```{r} -->
<!-- ## ------------- stage -->
<!-- axes <- colnames(all_pls_umap)[1:2] -->
<!-- gg_stage <- ggplot(all_pls_umap) +  -->
<!--   theme_classic() + -->
<!--   geom_point(aes_string(x = axes[1], y = axes[2], col = 'stage'), alpha = params$ggplot_alph) + -->
<!--   facet_wrap(.~Modality, ncol = 2, scales = 'free') + -->
<!--   scale_color_manual(values = col_palette) + -->
<!--   labs(col = 'Stage') -->
<!-- ## ------------- lineage -->
<!-- # lineage_col_palette <- viridisLite::viridis(n=length(unique(unique(gastru.mae$lineage)))) -->
<!-- lineage_col_palette <- c("#440154FF", "#3B528BFF", "#21908CFF", "#5DC863FF", "#FDE725FF") -->
<!-- names(lineage_col_palette) <- unique(gastru.mae$lineage) -->
<!-- axes <- colnames(all_pls_umap)[1:2] -->
<!-- gg_lineage <- ggplot(all_pls_umap) +  -->
<!--   theme_classic() + -->
<!--   geom_point(aes_string(x = axes[1], y = axes[2], col = 'lineage'), alpha = params$ggplot_alph) + -->
<!--   facet_wrap(.~Modality, ncol = 2, scales = 'free') + -->
<!--   scale_color_manual(values = lineage_col_palette) + -->
<!--   labs(col = 'Putative Lineage') -->
<!-- ``` -->
<!-- ```{r, fig.width=12, fig.height=6} -->
<!-- gg_stage_lineage <- gridExtra::grid.arrange(gg_stage + labs(x = ''), -->
<!--                         ggplot() + theme_minimal() + labs(x = 'UMAP_1'),  -->
<!--                         gg_lineage + labs(y = '', x = ''),  -->
<!--                         widths = c(10,1,10),  -->
<!--                         nrow=1) -->
<!-- ``` -->
<!-- UMAP projection using 4 PLS components for each modality shows the distinct states cells from each stage on transcriptional and DNA methylation level in the projected space. This suggests that the molecular changes during mouse gastrulation are coordinated across the study modalities. The projection also highlights the different patterns of transcriptional variation among different putative lineages in late-stage cells. -->
<!-- ```{r, eval=params$save_output, echo=FALSE} -->
<!-- ggsave(plot = gg_stage_lineage, filename = 'figures/UMAP-MultiModalSparsePLS-stage-lineage.pdf', width = 12, height = 6) -->
<!-- ggsave(plot = gg_stage, filename = 'figures/UMAP-MultiModalSparsePLS-stage.pdf', width = 10, height = 9) -->
<!-- ``` -->
<!-- ## concordance with RNA -->
<!-- Here we calculate the correlations between components from gene expression and different methylome modalities as a measure of concordnat variation: -->
<!-- ```{r} -->
<!-- ## For a pls object which includes pls cell embeddings from a computed latent space -->
<!-- ## where covariance with rna is maximal, calculate the correlation b/w rna scores -->
<!-- ## and those of a given modality's (per component and average) - potentially weighted by amount of  -->
<!-- ## variance each rna component explains (weighted=TRUE) - as a measure of concordant variation with rna -->
<!-- concordance <- function(pls_obj,  -->
<!--                         comp = 1, ## which componets to use -->
<!--                         weighted = TRUE,  -->
<!--                         which.plot = NULL, ## which component to plot (or [weighted.]mean) -->
<!--                         show.legend = TRUE -->
<!--                         ) { -->
<!--   if (!is.null(which.plot)) { -->
<!--     which.plot[is.numeric(which.plot)] <- paste0('Component ',  which.plot[is.numeric(which.plot)]) -->
<!--   } -->
<!--   ## correlation of components with RNA components -->
<!--   cor_with_y <- sapply(pls_obj$variates[-1], function(x_variates) { -->
<!--     y_variates <- pls_obj$variates[[1]] -->
<!--     y_variates <- y_variates[, comp, drop = FALSE] -->
<!--     cor <- cor(y_variates, x_variates) -->
<!--     cor <- diag(cor) -->
<!--     names(cor) <- paste0('Component ', comp) -->
<!--     if (isFALSE(weighted)) { -->
<!--       weight <- rep(1, length(cor)) -->
<!--     } else { -->
<!--       weight <- pls_obj$explained_variance$rna -->
<!--     } -->
<!--     cor.mean <- mean(cor) -->
<!--     wtd.cor.mean <- weighted.mean(x = cor, w = weight) -->
<!--     cor['mean'] <- cor.mean -->
<!--     cor['weighted.mean'] <- wtd.cor.mean -->
<!--     cor <- round(cor, 2) -->
<!--     cor -->
<!--   }) -->
<!--   print(kable(cor_with_y)) -->
<!--   ## make a long data.frame -->
<!--   cor_with_y <- cor_with_y %>% as.data.frame() %>% t() %>%  -->
<!--     reshape2::melt(id.vars = 1:3) %>%  -->
<!--     set_colnames(c('Modality', 'Component', 'Concordance')) -->
<!--   if (!is.null(which.plot)) { -->
<!--     cor_with_y <- cor_with_y[cor_with_y$Component %in% which.plot,] -->
<!--   } -->
<!--   ggplot(cor_with_y, aes(x = Modality, y = Concordance)) + geom_col(aes(fill = Component), position = 'dodge2', show.legend = show.legend) + theme_classic() + -->
<!--     labs(y = sprintf('Correlation with RNA (comp %s) ', paste(comp, collapse = ', ')), x = '', fill = '') + ylim(c(0, 1)) + theme(axis.text.x = element_text(angle = 45, vjust = 0.8, hjust=1)) -->
<!-- } -->
<!-- ``` -->
<!-- ```{r} -->
<!-- concordance(pls_obj = mmspls, comp = c(1:4), which.plot = NULL) -->
<!-- ``` -->
</div>
</div>
<div id="imputation" class="section level1">
<h1 class="hasAnchor">
<a href="#imputation" class="anchor"></a>Imputation</h1>
<p>Impute the missing methylation data using a nearest neighbours algorithm:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># sum(!gastru.mae$pass_metQC)</span>
<span class="co">#&gt; 0</span>
<span class="va">gastru.mae.imputed</span> <span class="op">&lt;-</span> <span class="va">gastru.mae</span>
<span class="va">met_assays</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">'^met'</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment-methods.html">experiments</a></span><span class="op">(</span><span class="va">gastru.mae.imputed</span><span class="op">)</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">assay_type</span> <span class="kw">in</span> <span class="va">met_assays</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">assay_values</span> <span class="op">&lt;-</span> <span class="va">gastru.mae.imputed</span><span class="op">@</span><span class="va">ExperimentList</span><span class="op">[[</span><span class="va">assay_type</span><span class="op">]</span><span class="op">]</span>
  <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">'\nimputing values for:'</span>, <span class="va">assay_type</span>, <span class="st">'\n'</span><span class="op">)</span>
  <span class="va">gastru.mae.imputed</span><span class="op">@</span><span class="va">ExperimentList</span><span class="op">[[</span><span class="va">assay_type</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> 
    <span class="fu"><a href="https://rdrr.io/pkg/impute/man/impute.knn.html">impute.knn</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">assay_values</span>, 
               k <span class="op">=</span> <span class="fl">30</span>, 
               rowmax <span class="op">=</span> <span class="fl">0.99</span>, 
               colmax <span class="op">=</span> <span class="fl">0.99</span>,  
               rng.seed<span class="op">=</span><span class="fl">42</span><span class="op">)</span><span class="op">$</span><span class="va">data</span>
<span class="op">}</span></code></pre></div>
<div id="data-integration-using-imputed-methylation-data" class="section level2">
<h2 class="hasAnchor">
<a href="#data-integration-using-imputed-methylation-data" class="anchor"></a>Data integration using imputed methylation data:</h2>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span><span class="op">(</span><span class="st">'Running PLS on imputed data with %s components performing variable selection\n'</span>, <span class="va">ncomp</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Running PLS on imputed data with 4 components performing variable selection</span>
<span class="va">st</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span><span class="op">(</span><span class="op">{</span>
  <span class="va">mmspls.imputed</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="../reference/multimodal_sPLS_wrapper.html">multimodal_sPLS_wrapper</a></span><span class="op">(</span>mae <span class="op">=</span> <span class="va">gastru.mae.imputed</span>, study_assays <span class="op">=</span> <span class="cn">NULL</span>,
                            ncomp <span class="op">=</span> <span class="va">ncomp</span>, scale <span class="op">=</span> <span class="va">scale</span>, design <span class="op">=</span> <span class="st">'null'</span>, lineages <span class="op">=</span> <span class="cn">NULL</span>,
                            stages <span class="op">=</span> <span class="cn">NULL</span>, DA <span class="op">=</span> <span class="cn">NULL</span>, keepX <span class="op">=</span> <span class="cn">NULL</span>, save <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span><span class="op">[</span><span class="st">'elapsed'</span><span class="op">]</span>
<span class="co">#&gt; Making unique feature names by adding assay name to feature name for: met_genebody</span>
<span class="co">#&gt; Making unique feature names by adding assay name to feature name for: met_promoter</span>

<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">'\nPLS run finished. Runtime: '</span>, <span class="fu">format_runtime</span><span class="op">(</span><span class="va">st</span><span class="op">)</span>, <span class="st">'\n'</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; PLS run finished. Runtime:  00:00:39</span></code></pre></div>
</div>
<div id="score-plots-2" class="section level2">
<h2 class="hasAnchor">
<a href="#score-plots-2" class="anchor"></a>score plots</h2>
<p>Plot PLS components for cells across all views:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pls_impute</span> <span class="op">&lt;-</span> <span class="fu">plot_pls</span><span class="op">(</span>pls_obj <span class="op">=</span> <span class="va">mmspls.imputed</span>, stage <span class="op">=</span> <span class="va">gastru.mae.imputed</span><span class="op">$</span><span class="va">stage</span>, col_palette <span class="op">=</span> <span class="va">col_palette</span><span class="op">)</span>
<span class="va">pls_impute</span></code></pre></div>
<p><img src="index_files/figure-html/unnamed-chunk-24-1.png" width="768"></p>
<p>The met_DHS view has clearly improved in separating different stages.</p>
<!-- ## UMAP using 4 PLS components from imputed data -->
<!-- ```{r} -->
<!-- for (block in names(mmspls.imputed$variates)) { -->
<!--   variates <- mmspls.imputed$variates[block] -->
<!--   dimred.name <- paste0('PLS_imputed_', block) -->
<!--   umap.name <- paste0('UMAP_PLS_imputed_', block) -->
<!--   reducedDim(rna.sce.matching, dimred.name) <- mmspls.imputed$variates[[block]] -->
<!--   set.seed(params$umap_params['run.seed']) -->
<!--   rna.sce.matching <- runUMAP(rna.sce.matching, dimred=dimred.name,  -->
<!--                               ncomponents = params$umap_params['n_components'],  -->
<!--                               n_neighbors = params$umap_params['n_neighbors'],  -->
<!--                               min_dist = params$umap_params['min_dist'],  -->
<!--                               name = umap.name) -->
<!-- } -->
<!-- ``` -->
<!-- Create a long data.frame of all variates for facet plotting:  -->
<!-- ```{r} -->
<!-- all_pls_umap <- list() -->
<!-- for (block in names(mmspls.imputed$variates)) { -->
<!--   umap.name <- paste0('UMAP_PLS_imputed_', block) -->
<!--   umap_df <- as.data.frame(reducedDim(rna.sce.matching, umap.name)) -->
<!--   colnames(umap_df) <- paste0('UMAP_', seq_along(umap_df)) -->
<!--   umap_df$stage <- gastru.mae.imputed$stage -->
<!--   umap_df$lineage <- gastru.mae.imputed$lineage -->
<!--   all_pls_umap[[block]] <- umap_df -->
<!-- } -->
<!-- all_pls_umap <- rbindListWithNames(all_pls_umap, new_col = 'Modality') -->
<!-- ``` -->
<!-- Plot UMAP: -->
<!-- ```{r} -->
<!-- ## ------------- stage -->
<!-- axes <- colnames(all_pls_umap)[1:2] -->
<!-- gg_stage <- ggplot(all_pls_umap) +  -->
<!--   theme_classic() + -->
<!--   geom_point(aes_string(x = axes[1], y = axes[2], col = 'stage'), alpha = params$ggplot_alph) + -->
<!--   facet_wrap(.~Modality, ncol = 2, scales = 'free') + -->
<!--   scale_color_manual(values = col_palette) + -->
<!--   labs(col = 'Stage') -->
<!-- ## ------------- lineage -->
<!-- # lineage_col_palette <- viridisLite::viridis(n=length(unique(unique(gastru.mae.imputed$lineage)))) -->
<!-- lineage_col_palette <- c("#440154FF", "#3B528BFF", "#21908CFF", "#5DC863FF", "#FDE725FF") -->
<!-- names(lineage_col_palette) <- unique(gastru.mae.imputed$lineage) -->
<!-- axes <- colnames(all_pls_umap)[1:2] -->
<!-- gg_lineage <- ggplot(all_pls_umap) +  -->
<!--   theme_classic() + -->
<!--   geom_point(aes_string(x = axes[1], y = axes[2], col = 'lineage'), alpha = params$ggplot_alph) + -->
<!--   facet_wrap(.~Modality, ncol = 2, scales = 'free') + -->
<!--   scale_color_manual(values = lineage_col_palette) + -->
<!--   labs(col = 'Putative Lineage') -->
<!-- ``` -->
<!-- ```{r, fig.width=12, fig.height=6} -->
<!-- gg_stage_lineage <- gridExtra::grid.arrange(gg_stage + labs(x = ''), -->
<!--                         ggplot() + theme_minimal() + labs(x = 'UMAP_1'),  -->
<!--                         gg_lineage + labs(y = '', x = ''),  -->
<!--                         widths = c(10,1,10),  -->
<!--                         nrow=1) -->
<!-- ``` -->
<!-- UMAP projection using imputed data: -->
<!-- ```{r, eval=params$save_output, echo=FALSE} -->
<!-- ggsave(plot = gg_stage_lineage, filename = 'figures/UMAP-MultiModalSparsePLS-imputed-stage-lineage.pdf', width = 12, height = 6) -->
<!-- ggsave(plot = gg_stage, filename = 'figures/UMAP-MultiModalSparsePLS-imputed-stage.pdf', width = 10, height = 9) -->
<!-- ``` -->
</div>
<div id="predictive-performance-of-pls-and-pca-components" class="section level2">
<h2 class="hasAnchor">
<a href="#predictive-performance-of-pls-and-pca-components" class="anchor"></a>Predictive performance of PLS and PCA components</h2>
<p>We finally calculate the balanced prediction accuracy of a kmeans clustering algorithm using the first 2 components from PCA components from individual assays as well as using PLS variates from integration of original and imputed measurements:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rna_pca</span> <span class="op">&lt;-</span> <span class="va">rna.sce.matching</span><span class="op">@</span><span class="va">int_colData</span><span class="op">$</span><span class="va">reducedDims</span><span class="op">@</span><span class="va">listData</span><span class="op">$</span><span class="va">PCA</span>
<span class="va">all_scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>rna <span class="op">=</span> <span class="va">rna_pca</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">nipals_comps</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">$</span><span class="va">scores</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">rna_pca</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="co"># rna_pca &lt;- rna_pca[rownames(mmspls$variates$rna),]</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">427</span><span class="op">)</span>
<span class="co">## kmeans clustering balanced accuracy with PCA components from RNA and DNA data:</span>
<span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
  PCA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">all_scores</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span>
        <span class="fu">kmeans_accuracy</span><span class="op">(</span>variates <span class="op">=</span> <span class="va">x</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, labels <span class="op">=</span> <span class="va">mmspls</span><span class="op">$</span><span class="va">stage</span><span class="op">)</span><span class="op">$</span><span class="va">BAR</span>
    <span class="op">}</span><span class="op">)</span>,
  <span class="co">## without imputation</span>
    PLS <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">mmspls</span><span class="op">$</span><span class="va">variates</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span>
        <span class="fu">kmeans_accuracy</span><span class="op">(</span>variates <span class="op">=</span> <span class="va">x</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, labels <span class="op">=</span> <span class="va">mmspls</span><span class="op">$</span><span class="va">stage</span><span class="op">)</span><span class="op">$</span><span class="va">BAR</span>
    <span class="op">}</span><span class="op">)</span>,
    <span class="co">## with imputation</span>
    PLS_impute <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">mmspls.imputed</span><span class="op">$</span><span class="va">variates</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span>
        <span class="fu">kmeans_accuracy</span><span class="op">(</span>variates <span class="op">=</span> <span class="va">x</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, labels <span class="op">=</span> <span class="va">mmspls</span><span class="op">$</span><span class="va">stage</span><span class="op">)</span><span class="op">$</span><span class="va">BAR</span>
    <span class="op">}</span><span class="op">)</span>
<span class="op">)</span>

<span class="fu">kable</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">rna</th>
<th align="right">met_genebody</th>
<th align="right">met_promoter</th>
<th align="right">met_cgi</th>
<th align="right">met_DHS</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">PCA</td>
<td align="right">0.55</td>
<td align="right">0.76</td>
<td align="right">0.90</td>
<td align="right">0.88</td>
<td align="right">0.91</td>
</tr>
<tr class="even">
<td align="left">PLS</td>
<td align="right">0.88</td>
<td align="right">0.69</td>
<td align="right">0.86</td>
<td align="right">0.86</td>
<td align="right">0.88</td>
</tr>
<tr class="odd">
<td align="left">PLS_impute</td>
<td align="right">0.87</td>
<td align="right">0.81</td>
<td align="right">0.83</td>
<td align="right">0.91</td>
<td align="right">0.88</td>
</tr>
</tbody>
</table>
<p>Surprisingly, For normalised gene expression data, the PLS components are more predictive of the embryonic stage than PCA components. Noticeably, without imputation, the components from all assays are nearly equally predictive with the exception of genebody methylome which underperforms the rest. After imputation, the accuracy of the prediction using PLS components is improved considerably for genebody methylation.</p>
<p>Looking at individual stages:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## helpers and function to create a data.frame of form:</span>
<span class="co">##         Stage Method Modality Accuracy</span>
<span class="co">##          E4.5    PCA      rna     0.50</span>
<span class="co">##          E5.5    PLS      rna     0.35</span>
<span class="co">##          E6.5    PCA      rna     0.75</span>
<span class="va">.get_class_accuracy</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">scores</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">DF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">scores</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu">kmeans_accuracy</span><span class="op">(</span>variates <span class="op">=</span> <span class="va">x</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, labels <span class="op">=</span> <span class="va">mmspls</span><span class="op">$</span><span class="va">stage</span><span class="op">)</span><span class="op">$</span><span class="va">classAccuracy</span>
<span class="op">}</span><span class="op">)</span>
  <span class="va">DF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">DF</span><span class="op">)</span>,<span class="fl">2</span><span class="op">)</span>
  <span class="va">DF</span><span class="op">$</span><span class="va">Stage</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">DF</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">DF</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>
  <span class="va">DF</span>
<span class="op">}</span>

<span class="va">get_class_accuracy</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">scores_list</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">df_class</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">scores_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">scores</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu">.get_class_accuracy</span><span class="op">(</span><span class="va">scores</span><span class="op">)</span>
  <span class="op">}</span><span class="op">)</span>
  <span class="va">df_class</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/utils.html">rbindListWithNames</a></span><span class="op">(</span><span class="va">df_class</span>, new_col <span class="op">=</span> <span class="st">'Method'</span><span class="op">)</span>
  <span class="va">df_class</span> <span class="op">&lt;-</span> <span class="fu">reshape2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html">melt</a></span><span class="op">(</span><span class="va">df_class</span>, measure.vars <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, 
                             value.name <span class="op">=</span> <span class="st">'Accuracy'</span>, 
                             variable.name <span class="op">=</span> <span class="st">'Modality'</span><span class="op">)</span>
  <span class="va">df_class</span>
<span class="op">}</span></code></pre></div>
<p>Get per class accuracy:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">accuracy_per_class</span> <span class="op">&lt;-</span> 
  <span class="fu">get_class_accuracy</span><span class="op">(</span>scores_list <span class="op">=</span> 
                       <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>PCA <span class="op">=</span> <span class="va">all_scores</span>, 
                            PLS <span class="op">=</span> <span class="va">mmspls</span><span class="op">$</span><span class="va">variates</span>,
                            PLS_impute <span class="op">=</span> <span class="va">mmspls.imputed</span><span class="op">$</span><span class="va">variates</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">accuracy_per_class</span><span class="op">)</span>
<span class="co">#&gt;   Stage Method Modality Accuracy</span>
<span class="co">#&gt; 1  E4.5    PCA      rna     0.50</span>
<span class="co">#&gt; 2  E5.5    PCA      rna     0.49</span>
<span class="co">#&gt; 3  E6.5    PCA      rna     0.58</span>
<span class="co">#&gt; 4  E7.5    PCA      rna     0.64</span>
<span class="co">#&gt; 5  E4.5    PLS      rna     1.00</span>
<span class="co">#&gt; 6  E5.5    PLS      rna     0.99</span></code></pre></div>
<p>Column plots of balanced accuracy rates for all modalities and stages:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">accuracy_per_class</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">Stage</span>, <span class="va">Accuracy</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="va">.</span><span class="op">~</span><span class="va">Modality</span>, ncol<span class="op">=</span><span class="fl">2</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_col</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">Method</span><span class="op">)</span>, position <span class="op">=</span> <span class="st">'dodge'</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">'Balanced Accuracy'</span><span class="op">)</span></code></pre></div>
<p><img src="index_files/figure-html/unnamed-chunk-29-1.png" width="768"></p>
<p>Views differ in their predictive performance for different stages. For instance, using PLS components the genebody methylation more predictive of embryonic stages in E4.5 and E5.5 cells compared to CGI methylation, whereas the opposite holds for late-stage cells.</p>
</div>
</div>
<div id="conclusion" class="section level1">
<h1 class="hasAnchor">
<a href="#conclusion" class="anchor"></a>Conclusion</h1>
<p>Multi-modal PLS derives the common variation between transcriptome and methylome in different genomic regions for 795 single cells during mouse gastrulation. The used data types vary in dimensions and and biological variation. The common axes of variation are driven partly by the stages of embryo, which mainly separates early-stage and late-stage cells. The measurements include varying levels of masked values which are ignored by PLS algorithm during integration. Imputation of missing values using a nearest neighbours imputation method enabled the method to capture more cross-modality covariation relevant to embryonic stages in some data views. A classification of cells based on either PCA or PLS components revealed that the projected components of shared variation are more predictive of the embryonic stages than the Principal Components of the transcriptome. It also showed that different views resolve different phenotypic dynamics which highlights the importance of data integration to better understand the cellular behaviour.</p>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>Melbourne Integrative Genomics, The University of Melbourne (<a href="mailto:al.jal.abadi@gmail.com" class="email">al.jal.abadi@gmail.com</a>)<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Al JalalAbadi.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
